File: sdk/dart/api/lib/base/zlg_base_module.dart
import 'package:flutter/material.dart';

import 'client_error_code.dart';
import '../framework/zlg_http.dart';

class ZLGBaseModule {
  ValueChanged<int> errorHandler;

  void init({String baseUrl, int timeout}) {
    ZlgHttp.getInstance(baseUrl: baseUrl, timeout: timeout);
  }

  Future<Response<T>> fetch<T>(
      {String url,
      String method = 'get',
      data,
      Map<String, dynamic> params}) async {
    return ZlgHttp.getInstance()
        .fetch<T>(url: url, method: method, data: data, params: params);
  }

  void setErrorHandler(ValueChanged<int> handler) {
    errorHandler = handler;
  }

  void handleError(Map<int, int> errorMap, Response errResponse) {
    print('handleError err: $errResponse');
    int status, code;
    if(errResponse == null) {
      status = 1000;
    } else {
      status = errResponse.statusCode;
      code = errResponse.data['code'];
    }
    var commonErrorMap = {
      400: ClientErrorCode.ZLG_PARAM_WRONG,
      401: ClientErrorCode.ZLG_NOT_LOGIN,
      403: ClientErrorCode.ZLG_FORBIDDEN,
      408: ClientErrorCode.ZLG_REQUEST_TIMEOUT,
      500: ClientErrorCode.ZLG_SERVER_ERROR,
      1000: ClientErrorCode.ZLG_NO_NETWORK
    };
    var errorCode = code ??
        errorMap[status] ??
        commonErrorMap[status] ??
        ClientErrorCode.ZLG_UNKNOW;
    if (errorHandler != null) {
      errorHandler(errorCode);
    } else {
      throw {'errorCode': errorCode};
    }
  }
}

File: sdk/js/api/http/httpBase/ZLGBaseModule.js
import ZLGHttp from './ZLGHttp'
import ZLGHttpClientCode from './ZLGHttpClientCode'
import ZLGHttpErrors from './ZLGHttpErrors'

class ZLGBaseModule {
  static init (baseURL, timeout, header) {
    ZLGHttp.getInstance(baseURL, timeout, header)
  }

  static fetch (params) {
    const http = ZLGHttp.getInstance()
    return http.fetch(params)
  }

  static setErrorHandler (handler) {
    if (typeof handler === 'function') {
      this.errorHandler = handler
    } else {
      throw new Error('Error Handler Must Be Function')
    }
  }

  static handleError (errorMap, err) {
    let status = err.status
    let code = err.data.code
    const commonErrorMap = {
      400: ZLGHttpClientCode.ZLG_PARAM_WRONG,
      401: ZLGHttpClientCode.ZLG_NOT_LOGIN,
      403: ZLGHttpClientCode.ZLG_FORBIDDEN,
      408: ZLGHttpClientCode.ZLG_REQUEST_TIMEOUT,
      500: ZLGHttpClientCode.ZLG_SERVER_ERROR,
      NO_RESPONSE: ZLGHttpClientCode.ZLG_SERVER_NO_RESPONSE
    }
    let errorCode = code || errorMap[status] || commonErrorMap[status] || ZLGHttpClientCode.ZLG_UNKNOW
    if (!this.errorHandler || !this.errorHandler(ZLGHttpErrors[errorCode] || errorCode)) {
      /* eslint-disable-next-line */
      return Promise.reject({
        errorCode: ZLGHttpErrors[errorCode] || errorCode
      })
    }
  }
}

export default ZLGBaseModule
